src/System.Console/src/System/ConsolePal.Unix.cs

```diff
@@ -49,6 +52,75 @@ public static Encoding OutputEncoding
             get { return GetConsoleEncoding(); }
         }
 
+        private static readonly object InternalSyncObject = new object();
+        private static SyncTextReader s_stdInReader;
+        private const int DefaultBufferSize = 255;
+
+        public static SyncTextReader StdInReader
+        {
+            get
+            {
+                if (Console.IsInputRedirected)
```

Comment:
This check should probably be moved down to the result == null if block.  For cases where input isn't redirected, we don't need to be checking this every time we access StdInReader, as all but the first time it'll have been initialized and redirection won't change.  For cases where input is redirected, we're going to throw anyway, so it doesn't matter if we do a little bit of unnecessary work first.